src/main/java/org/breedinginsight/brapps/importer/services/processors/experiment/middleware/Middleware.java

```diff
@@ -0,0 +1,60 @@
+package org.breedinginsight.brapps.importer.services.processors.experiment.middleware;
+
+import org.breedinginsight.brapps.importer.services.processors.experiment.model.MiddlewareError;
+
+public abstract class Middleware<T> {
+
+    Middleware next;
```

Comment:
The processors already consist of four main tasks---initialize, process, validate, and post---done in a specific order so the list structure is natural fit. The steps can be as granular as is necessary, so I put things together in the same step if they belong conceptually but separate out if there are associated undo operations. For example, I put all the initialization logic in a single middleware and the processing in a single middleware, but the posting of data has multiple middleware steps since there is a an undo operation for these steps.

This works for now, but I think in the future, a better way along the lines you suggest would be to use an event-based observer pattern instead of middleware for the posting of brapi data, so the order of the operations doesn't have to be written out and maintained explicitly.